// @serenity-js/webdriverio/esm/screenplay/models/WebdriverIOModalDialogHandler.js
import 'webdriverio';
import { AbsentModalDialog, AcceptedModalDialog, DismissedModalDialog, ModalDialogHandler } from '@serenity-js/web';
import { WebdriverProtocolErrorCode } from './WebdriverProtocolErrorCode.js';
/**
 * WebdriverIO-specific implementation of [`ModalDialogHandler`](https://serenity-js.org/api/web/class/ModalDialogHandler/),
 * used with the [WebDriver protocol](https://webdriver.io/docs/api/webdriver).
 *
 * @group Models
 */
export class WebdriverIOModalDialogHandler extends ModalDialogHandler {
    browser;
    defaultHandler = async (dialog) => {
        const message = dialog.message();
        await dialog.dismiss();
        this.modalDialog = new DismissedModalDialog(message);
    };
    currentHandler;
    dialog;
    constructor(browser) {
        super();
        this.browser = browser;
        this.currentHandler = this.defaultHandler;
        this.browser.on('dialog', this.onDialog);
    }
    onDialog = this.tryToHandleDialog.bind(this);
    async tryToHandleDialog(dialog) {
        try {
            await this.currentHandler(dialog);
        }
        catch (error) {
            if (error.name === WebdriverProtocolErrorCode.UnexpectedAlertOpenError) {
                await dialog.dismiss();
                return;
            }
            if (error.message.includes(WebdriverProtocolErrorCode.NoSuchAlertError)) {
                this.modalDialog = new AbsentModalDialog();
                return;
            }
            throw error;
        }
    }
    async acceptNext() {
        this.currentHandler = async (dialog) => {
            const message = dialog.message();
            await this.browser.acceptAlert();
            this.modalDialog = new AcceptedModalDialog(message);
        };
    }
    async acceptNextWithValue(text) {
        this.currentHandler = async (dialog) => {
            const message = dialog.message();
            await dialog.accept(String(text));
            this.modalDialog = new AcceptedModalDialog(message);
        };
    }
    async dismissNext() {
        this.currentHandler = async (dialog) => {
            const message = dialog.message();
            await dialog.dismiss();
            this.modalDialog = new DismissedModalDialog(message);
        };
    }
    async dismiss() {
        if (!this.dialog) {
            return;
        }
        const message = this.dialog.message();
        await this.dialog.dismiss();
        this.modalDialog = new DismissedModalDialog(message);
    }
    async reset() {
        this.modalDialog = new AbsentModalDialog();
        this.currentHandler = this.defaultHandler;
    }
    /**
     * @override
     */
    async last() {
        return this.modalDialog;
    }
    async discard() {
        this.browser.off('dialog', this.onDialog);
    }
}
